src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -1481,5 +1481,251 @@ public static double ScaleB(double x, int n)
+            return 0;
+        }
+
+        [StackTraceHidden]
+        private static uint DoubleToUIntOverflow(double val)
+        {
+            // Note that this expression also works properly for val = NaN case
+            if (val is > -1.0 and < 4294967296.0)
+            {
+                uint ret = (uint)(long)val;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(val - 1.0 <= ret);
+                Debug.Assert(ret <= val + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [StackTraceHidden]
+        private static long DoubleToLongOverflow(double val)
+        {
+            const double two63 = 2147483648.0 * 4294967296.0;
+
+            // Note that this expression also works properly for val = NaN case
+            // We need to compare with the very next double to two63. 0x402 is epsilon to get us there.
+            if (val is > -two63 - 0x402 and < two63)
+            {
+                long ret = (long)val;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(val - 1.0 <= ret);
+                Debug.Assert(ret <= val + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [StackTraceHidden]
+        private static ulong DoubleToULongOverflow(double val)
+        {
+            const double two64 = 4294967296.0 * 4294967296.0;
+            // Note that this expression also works properly for val = NaN case
+            if (val is > -1.0 and < two64)
+            {
+                ulong ret = (ulong)val;
+                // since no overflow can occur, the value always has to be within 1
+                Debug.Assert(val - 1.0 <= ret);
+                Debug.Assert(ret <= val + 1.0);
+                return ret;
+            }
+
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        private static double DoubleReminder(double dividend, double divisor)
+        {
+            // From the ECMA standard:
+            //
+            // If [divisor] is zero or [dividend] is infinity
+            //   the result is NaN.
+            // If [divisor] is infinity,
+            //   the result is [dividend] (negated for -infinity***).
+            //
+            // ***"negated for -infinity" has been removed from the spec
+            if (divisor == 0 || !double.IsFinite(dividend))
+            {
+                return double.NaN;
+            }
+
+            if (!double.IsFinite(divisor) && !double.IsNaN(divisor))
+            {
+                return dividend;
+            }
+
+            return FMod(dividend, divisor);
+        }
+
+        private static int DoubleToInt(double val) => (int)(long)val;
+        private static uint DoubleToUInt(double val) => (uint)(long)val;
```

Comment:
Do we really need helpers this simple? Is it unsuitable for JIT to expand one cast into two?

Reply:
Those are needed for existing R2R binaries and can be removed after a major R2R version bump.